sourced/ml/mining/cmd/collect_stdlibs.py

```diff
@@ -0,0 +1,98 @@
+    JavaStdlibSpider,
+    PythonStdlibSpider,
+    RubyStdlibSpider,
+)
+from sourced.ml.mining.utils import check_remove_filepath, path_with_suffix
+
+
+class StdlibPipeline(object):
+    """Item pipeline that holds data in memory then saves it as a StandardLibraries model."""
+
+    libraries = {}
+    infos = {}
+    not_done = []
+
+    def __init__(self, output_path):
+        """Create the item pipeline."""
+        self.output_path = output_path
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        """Pass the output path to the constructor when instanciating from the crawler."""
+        return cls(output_path=crawler.settings["OUTPUT_PATH"])
+
+    def open_spider(self, spider):
+        """Open a spider."""
+        self.not_done.append(None)
+        log = logging.getLogger(spider.name)
+        log.info("Opened spider")
+
+    def close_spider(self, spider):
+        """Close a spider, and save the model if it is the last to terminate."""
+        log = logging.getLogger(spider.name)
+        log.info("Closed spider")
+        self.not_done.pop()
+        if self.not_done:
+            n = len(self.not_done)
+            log.info("Still %d spider%s running" % (n, "s" * (n > 1)))
+        else:
+            log.info("No more spiders running, creating the model ...")
+            model = StandardLibraries(log_level=logging.INFO).construct(
+                self.libraries, self.infos
+            )
+            model.save(self.output_path, series="stdlib")
+            log.info("Saved model to %s" % self.output_path)
+
+    def process_item(self, item, spider):
+        """Process an item returned by one of the spiders."""
+        lang, library = item["lang"], item["library"]
+        if lang not in self.libraries:
+            self.libraries[lang] = set()
+            self.infos[lang] = {}
+        self.libraries[lang].add(library)
+        for info in item["info"]:
+            if info not in self.infos[lang]:
+                self.infos[lang][info] = set()
+            self.infos[lang][info].add(library)
+
+
+def collect_stdlibs(args: argparse.Namespace):
+    """
+    Collect the lists of standard libraries for each language Babelfish can parse.
+    """
+    log = logging.getLogger("collect_stdlibs")
+    output_path = path_with_suffix(args.output_path, ".asdf")
+    check_remove_filepath(output_path, log, args.force)
+    process = CrawlerProcess(
+        settings={
+            "BOT_NAME": "vegeta",
+            "ITEM_PIPELINES": {
+                "sourced.ml.mining.cmd.collect_stdlibs.StdlibPipeline": 100
+            },
+            "LOG_ENABLED": 0,
+            "OUTPUT_PATH": output_path,
+        }
+    )
+    logging.getLogger("scrapy").setLevel(logging.WARNING)
+
+    for spider in [
+        CppStdlibSpider,
+        CSharpStdlibSpider,
+        GoStdlibSpider,
+        JavaStdlibSpider,
+        PythonStdlibSpider,
+        RubyStdlibSpider,
+    ]:
+        process.crawl(spider)
+    process.start()
```

Comment:
Not so bad API after all. Scrapy should run your spiders all in parallel, in a nice Twisted reactor which is more efficient than threads.

Reply:
yeah, it's actually way faster now ! just a bit annoying to use when you're not directly calling `scrapy my_spider ...` and want to pass args, as well as keep items in memory when using multiple spiders